src/denoising_model.jl

```diff
@@ -0,0 +1,35 @@
+export denoising_model
+
+include("denoising_data.jl")
+
+function denoising_model(shape, shape_p, KERNEL_SIZE, KERNEL_TYPE, KERNEL_SIGMA = 1.5)
+  sigma = 10^-3
+  data_path = joinpath(@__DIR__, "..", "images/cameraman.png")
+  cameraman_image = load(data_path)
+  x_t = vec(Float64.(cameraman_image))
+  H, H_T, W, W_T = generate_HW(shape, shape_p, KERNEL_SIZE, KERNEL_TYPE, KERNEL_SIGMA)
+  (n, m) = shape
+  b = H(x_t) + randn(n * m) * sigma
+  y = similar(b)
+  z = similar(b)
+
+  function obj(x)
+    y .= W_T(x)
+    y .= H(y)
+    z .= log.((y - b) .^ 2 .+ 1)
+    return sum(z)
+  end
+
+  function grad!(g, x)
+    y = H(W_T(x))
+    term = 1.0 ./ ((y - b) .^ 2 .+ 1)
```

Comment:
`y .- b` ? Maybe you could also allocate `term` as you have done it with `b`, `y` and `z`?